emailproxy.py

```diff
@@ -337,9 +344,157 @@ def add_account(username):
+
+    @staticmethod
+    def _aws_secrets_fetch():
+        aws_client = AppConfig._aws_secrets_boto3_client()
+        if aws_client:
+            # get the values of all AWS Secrets referenced in the configuration file
+            aws_secrets = AppConfig._aws_secrets_ids()
+            for secret_id in aws_secrets:
+                try:
+                    get_secret_value_response = aws_client.get_secret_value(SecretId=secret_id)
+
+                except botocore.exceptions.ClientError as err_get_secret:
+                    if err_get_secret.response['Error']['Code'] == 'ResourceNotFoundException':
+                        Log.info('Fetching AWS Secret "%s"' % secret_id, '- secret does not exist or does not contain',
+                                 'a secret value')
+                    elif err_get_secret.response['Error']['Code'] == 'AccessDeniedException':
+                        Log.error('Fetching AWS Secret "%s"' % secret_id, '- access denied: does the IAM user have the',
+                                  '"secretsmanager:GetSecretValue" permission?')
+                    else:
+                        Log.error('Fetching AWS Secret "%s"' % secret_id, '- unexpected error; see the proxy\'s debug',
+                                  'mode log for details')
+                    AppConfig._aws_secrets_debug_log_boto3_error(err_get_secret)
+
+                else:
+                    aws_secrets[secret_id] = json.loads(get_secret_value_response['SecretString'])
+
+            # update local proxy configuration in memory
+            for account in AppConfig._aws_secrets_accounts():
+                account_secret = AppConfig._PARSER.get(account, 'aws_secret')
+                if account in aws_secrets[account_secret]:
+                    for key in AppConfig._SECRETS_MANAGER_KEYS:
+                        AppConfig._PARSER.set(account, key, aws_secrets[account_secret][account][key])
+        else:
+            Log.error('Unable to get AWS client; cannot fetch OAuth 2.0 tokens from AWS Secrets Manager')
+
+    @staticmethod
+    def _aws_secrets_save():
+        # a deep copy of the configuration file to be saved locally, *without* OAuth 2.0 tokens (_SECRETS_MANAGER_KEYS)
+        appconfig_to_save = configparser.ConfigParser()
+        appconfig_to_save.read_dict(AppConfig._PARSER)
+
+        # remove OAuth 2.0 tokens from each account, transferring them into a local dict of referenced AWS Secrets
+        aws_secrets = AppConfig._aws_secrets_ids()
+        for account in AppConfig._aws_secrets_accounts():
+            account_secret = appconfig_to_save.get(account, 'aws_secret')
+            aws_secrets[account_secret][account] = {key: appconfig_to_save.get(account, key) for key in
+                                                    AppConfig._SECRETS_MANAGER_KEYS}
+            for key in AppConfig._SECRETS_MANAGER_KEYS:
+                appconfig_to_save.remove_option(account, key)
```

Comment:
```suggestion
            aws_secrets[account_secret][account] = {}
            for key in [k for k in AppConfig._SECRETS_MANAGER_KEYS if k in appconfig_to_save[account]]:
                aws_secrets[account_secret][account][key] = appconfig_to_save.get(account, key)
                appconfig_to_save.remove_option(account, key)
```
Here's a potentially preferable solution, which avoids the use of an empty string to denote an undefined key.

Reply:
```suggestion
            aws_secrets[account_secret][account] = {}
            for key in AppConfig._SECRETS_MANAGER_KEYS:
                if appconfig_to_save.has_option(account, key):
                    aws_secrets[account_secret][account][key] = appconfig_to_save.get(account, key)
                    appconfig_to_save.remove_option(account, key)
```

I think I'd prefer using the configparser methods rather than direct dict access – here's an alternative approach

Reply:
Agreed your proposal is cleaner, added to commit e11338d